module/rdpCapture.c

```diff
@@ -819,35 +939,208 @@ rdpCapture2(rdpClientCon *clientCon, RegionPtr in_reg, BoxPtr *out_rects,
     int dst_stride;
     int crc_offset;
     int crc_stride;
-    int crc;
+    uint64_t crc;
     int num_crcs;
+    int num_skips;
+    int tile_row_stride, crc_height;
+    int i, best_offset, num_diff_first_rows;
+    int xn, px, yn, py;
+    int seed_y, found, tiles_scrolled;
+    uint64_t *row_hashes;
+    uint64_t old_first_row, new_first_row;
+    uint16_t *offset_histogram;
+    uint16_t max_count;
+    uint8_t *scrolled_map;
+    uint64_t old_hash, new_hash;
+    uint64_t target_hash;
+    BoxRec foundRect;
 
     LLOGLN(10, ("rdpCapture2:"));
-
-    *out_rects = g_new(BoxRec, RDP_MAX_TILES);
-    if (*out_rects == NULL)
-    {
-        return FALSE;
-    }
-    out_rect_index = 0;
+    extents_rect = *rdpRegionExtents(in_reg);
 
     src = id->pixels;
     dst = id->shmem_pixels;
     src_stride = id->lineBytes;
     dst_stride = clientCon->cap_stride_bytes;
 
     crc_stride = (clientCon->dev->width + 63) / 64;
-    num_crcs = crc_stride * ((clientCon->dev->height + 63) / 64);
+    /* tile rows are column-major */
+    tile_row_stride = ((clientCon->dev->height + 63) / 64) * 64;
+    crc_height = ((clientCon->dev->height + 63) / 64);
+    num_crcs = crc_stride * crc_height;
     if (num_crcs != clientCon->num_rfx_crcs_alloc)
     {
-        /* resize the crc list */
+        /* resize the hash list */
         clientCon->num_rfx_crcs_alloc = num_crcs;
         free(clientCon->rfx_crcs);
-        clientCon->rfx_crcs = g_new0(int, num_crcs);
+        free(clientCon->rfx_tile_row_hashes);
+        clientCon->rfx_crcs = g_new0(uint64_t, num_crcs);
+        clientCon->rfx_tile_row_hashes = g_new0(uint64_t, num_crcs * 64);
     }
 
-    extents_rect = *rdpRegionExtents(in_reg);
+    /* update the tile row hashes. Column major order to be kind to 
+       prefetchers even though it shouldn't matter much */
+    num_diff_first_rows = 0;
+    x = extents_rect.x1 & ~63;
+    while (x < extents_rect.x2)
+    {
+        y = extents_rect.y1 & ~63;
+        while (y < extents_rect.y2)
+        {
+            rect.x1 = x;
+            rect.y1 = y;
+            rect.x2 = rect.x1 + 64;
+            rect.y2 = rect.y1 + 64;
+            rcode = rdpRegionContainsRect(in_reg, &rect);
+            if (rcode == rgnIN)
+            {
+                row_hashes = clientCon->rfx_tile_row_hashes + 
+                    (x / 64) * tile_row_stride + y;
+                old_first_row = row_hashes[0];
+                wyhash_rfx_tile_rows(src, src_stride, x, y, row_hashes, 64);
+                new_first_row = row_hashes[0];
+                num_diff_first_rows += (old_first_row != new_first_row);
+            }
+            y += 64;
+        }
+        x += 64;
+    }
+
+    *scroll_offset = 0;
+    scroll_rect->x1 = 0;
+    scroll_rect->x2 = 0;
+    scroll_rect->y1 = 0;
+    scroll_rect->y2 = 0;
+
+    /* only try to detect a scroll rectangle if area is big enough */
+    if(extents_rect.y2 - extents_rect.y1 >= SCROLL_DET_MIN_H &&
```

Comment:
Alternatively if anyone just wants to quickly reap the benefits of this patch without caring about dead code, you can do what I did and edit the patch to put `0 &&` in this if statement.